src/nft/nft.cc

```diff
@@ -328,31 +329,28 @@ State Nft::insert_word(const State source, const Word &word, const State target)
     return word_tgt;
 }
 
-State Nft::insert_word(const State source, const Word &word) { return insert_word(source, word, add_state()); }
+State Nft::insert_word(const State source, const Word& word) {
+    assert(source < levels.size());
+    return insert_word(source, word, add_state_with_level(levels[source]));
+}
 
 State Nft::insert_word_by_parts(const State source, const std::vector<Word> &word_parts_on_levels, const State target) {
     assert(0 < num_of_levels);
     assert(word_parts_on_levels.size() == num_of_levels);
     assert(source < num_of_states());
     assert(target < num_of_states());
-    assert(levels[source] == 0);
-    assert(levels[target] == 0);
+    assert(source < levels.size());
```

Comment:
Isn't the assertion on `num_of_states()` redundant when `levels.size()` is used? Also, the assertion `assert(target < levels.size()` should be added.

Reply:
Theoretically, it is not redundant. One can extend, for example, delta without also extending `levels`. These asserts make sure that the sizes are synchronized as the algorithm utilizes these assumptions in the code.

The additional assert was added.

Reply:
I see. I would maybe add :
```cpp
size_t Nft::num_of_states() const {
    return std::max(Nfa::num_of_states(), levels.size());
}
```
into `Nft`. After that, `source < levels.size()` would imply `source < num_of_states()`.

Reply:
It would make sense if we agree that we can have a state which has a level, but is used nowhere else (not in `delta`, `initial` and `final`). As of now, however, it would break everything again since `num_of_states()` is used to get a new state, but I use `levels.resize()` in string operations to preallocate the levels for a predetermined number of states that will be created. I can make a note about this, and we can implement this later if we all agree that this semantics of `levels` is desired.

Reply:
Ok. Leave it as is for now.